File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> dq = new ArrayDeque<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (c == ')' || c == ']' || c == '}') {
                if (dq.size() == 0) {
                    return false;
                }
                
                char popped = dq.pop();
                
                if (c == ')' && popped != '(') {
                    return false;
                }
                                
                if (c == ']' && popped != '[') {
                    return false;
                }
                                
                if (c == '}' && popped != '{') {
                    return false;
                }
            } else {
                dq.push(c);
            }
        }
        
        return dq.size() == 0;
    }
}

File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for c in s:
            if c == ')' or c == '}' or c == ']':
                if len(stack) == 0:
                    return False
                
                popped = stack.pop()
                
                if c == ')' and popped != '(':
                    return False
                
                if c == '}' and popped != '{':
                    return False
                    
                if c == ']' and popped != '[':
                    return False
                
            else:
                stack.append(c)
        
        return len(stack) == 0